### PR TITLE
[AMDGPU] Enable atomic optimizer for 64 bit divergent values

### PR DESCRIPTION
**Pull request by vikramRH**
Kindly review only the top commit here, all the remaining changes are the same as in #89217 and #92725. 
This is the final patch in the series.

### CODE REVIEW

**Comment by arsenm** [comment]
> Kindly review only the top commit here

If you're going to repost with a pre-commit, it would be better to have all the pieces squashed into one. Also you could look into using graphite or SPR for managing dependent pull requests

File: llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp [hunk 1]
```diff
@@ -228,10 +228,11 @@ void AMDGPUAtomicOptimizerImpl::visitAtomicRMWInst(AtomicRMWInst &I) {
...
   // If the value operand is divergent, each lane is contributing a different
   // value to the atomic calculation. We can only optimize divergent values if
-  // we have DPP available on our subtarget, and the atomic operation is 32
-  // bits.
+  // we have DPP available on our subtarget, and the atomic operation is either
+  // 32 or 64 bits.
   if (ValDivergent &&
-      (!ST->hasDPP() || DL->getTypeSizeInBits(I.getType()) != 32)) {
+      (!ST->hasDPP() || (DL->getTypeSizeInBits(I.getType()) != 32 &&
+      DL->getTypeSizeInBits(I.getType()) != 64))) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp:235]: You should move the type logic into a separate predicate function. But also, you should base this on the actual type and not just the bitwidth. Checking the bitwidth will break in the future when atomics support more vector operations

File: llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp [hunk 2]
```diff
@@ -311,10 +312,11 @@ void AMDGPUAtomicOptimizerImpl::visitIntrinsicInst(IntrinsicInst &I) {
 
   // If the value operand is divergent, each lane is contributing a different
   // value to the atomic calculation. We can only optimize divergent values if
-  // we have DPP available on our subtarget, and the atomic operation is 32
-  // bits.
+  // we have DPP available on our subtarget, and the atomic operation is 32 or
+  // 64 bits.
   if (ValDivergent &&
-      (!ST->hasDPP() || DL->getTypeSizeInBits(I.getType()) != 32)) {
+      (!ST->hasDPP() || (DL->getTypeSizeInBits(I.getType()) != 32 &&
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp:318]: Ditto

File: llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp (hunk 1 above)
Comment (vikramRH) [reply llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp]: Done

**Comment by vikramRH** [comment]
Apologies for the commit spam here, graphite seems a good option now onwards. However all dependent patches have landed now, the diff here is now up to date.

File: llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp [hunk 3]
```diff
@@ -402,34 +413,30 @@ Value *AMDGPUAtomicOptimizerImpl::buildReduction(IRBuilder<> &B,
 
   // Reduce within each pair of rows (i.e. 32 lanes).
   assert(ST->hasPermLaneX16());
-  V = B.CreateBitCast(V, IntNTy);
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp:405]: Please submit an NFC cleanup patch that just removes unnecessary bitcasting, before adding support for new atomic operations.

File: llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp [hunk 4]
```diff
@@ -178,6 +178,20 @@ bool AMDGPUAtomicOptimizerImpl::run(Function &F) {
...
 
+static bool shouldOptimize(Type *Ty) {
+  switch (Ty->getTypeID()) {
+  case Type::FloatTyID:
+  case Type::DoubleTyID:
+    return true;
+  case Type::IntegerTyID: {
+    if (Ty->getIntegerBitWidth() == 32 || Ty->getIntegerBitWidth() == 64)
+      return true;
+  }
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp:189]: Don't forget pointers

File: llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp [hunk 5]
```diff
@@ -178,6 +178,20 @@ bool AMDGPUAtomicOptimizerImpl::run(Function &F) {
   return Changed;
 }
 
+static bool shouldOptimize(Type *Ty) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUAtomicOptimizer.cpp:181]: Better name that expresses why this type is handleable. 

Also in a follow up, really should cover the i16/half/bfloat and 2 x half, 2 x bfloat cases

**Comment by vikramRH** [comment]
closing this in favour of https://github.com/llvm/llvm-project/pull/96933 and https://github.com/llvm/llvm-project/pull/96934

**Pull request closed without merging**